metaflow/plugins/pypi/conda_environment.py

```diff
@@ -164,25 +189,54 @@ def _path(url, local_path):
             )
             for id_, packages, _, platform in results:
                 if id_ in dirty:
+                    cached = True
                     self.write_to_environment_manifest([id_, platform, type_], packages)
+            if cached:
+                self.logger(
+                    f"Cached packages in {time.time() - start_time:.2f}s!",
+                    fg="yellow",
+                    bold=False,
+                    indent=True,
+                    overwrite=True,
+                    animate=True,
+                )
 
+        storage = None
+        if self.datastore_type not in ["local"]:
+            # Initialize storage for caching if using a remote datastore
+            storage = self.datastore(_datastore_packageroot(self.datastore, echo))
+
+        self.logger(
+            "Bootstrapping virtual environment(s) ...", fg="magenta", bold=False
+        )
+        start_time = time.time()
         # First resolve environments through Conda, before PyPI.
-        self.logger("Bootstrapping virtual environment(s) ...")
         for solver in ["conda", "pypi"]:
             with ThreadPoolExecutor() as executor:
-                results = list(
-                    executor.map(lambda x: solve(*x, solver), environments(solver))
-                )
-            _ = list(map(lambda x: self.solvers[solver].download(*x), results))
-            with ThreadPoolExecutor() as executor:
-                _ = list(
-                    executor.map(lambda x: self.solvers[solver].create(*x), results)
-                )
-            if self.datastore_type not in ["local"]:
-                # Cache packages only when a remote datastore is in play.
-                storage = self.datastore(_datastore_packageroot(self.datastore, echo))
-                cache(storage, results, solver)
-        self.logger("Virtual environment(s) bootstrapped!")
+                # parallel solves
+                solves = [
+                    executor.submit(lambda x: solve(*x, solver), env)
+                    for env in environments(solver)
+                ]
+                for future in as_completed(solves):
+                    result = future.result()
+                    # sequential downloads
+                    self.solvers[solver].download(*result)
+                    # parallel creates
+                    executor.submit(self.solvers[solver].create, *result)
+                    if storage:
+                        # parallel cache
+                        executor.submit(cache, storage, [result], solver)
```

Comment:
pypi execution can start a lot earlier

Reply:
need to be careful with the timing though, as a lot of pypi steps rely on
`self.micromamba.path_to_environment(id_)` which can get stuck to `None` if called too early.